CRM/Core/Resources.php

```diff
@@ -351,16 +359,19 @@ protected function mergeSettings($settings, $additions) {
    * Helper fn for addSetting.
    * Render JavaScript variables for the global CRM object.
    *
+   * @param string $region
+   *   The region to render the settings in
+   *
    * @return string
    */
-  public function renderSetting($region = NULL) {
+  public function renderSetting($region) {
     // On a standard page request we construct the CRM object from scratch
-    if (($region === 'html-header') || !self::isAjaxMode()) {
-      $js = 'var CRM = ' . json_encode($this->getSettings()) . ';';
+    if (($region === 'html-header') || (($region !== 'billing-block') && !self::isAjaxMode())) {
+      $js = 'var CRM = ' . json_encode($this->getSettings($region)) . ';';
     }
     // For an ajax request we append to it
     else {
-      $js = 'CRM.$.extend(true, CRM, ' . json_encode($this->getSettings()) . ');';
+      $js = 'CRM.$.extend(true, CRM, ' . json_encode($this->getSettings($region)) . ');';
```

Comment:
@colemanw @mattwire  I know this is a bit of a pre-existing issue, but that conditional (`if... $region === 'html-header'  || ...`) is growing a bit magical. Would it be possible to either:

* Have one JS fragment that's more clever ("if `CRM` exists, then merge data -- else initialize anew.").
* Simplify the rule: the *only* time to use `var CRM=...` is in `html-header`. (If you use it at any other time, whether AJAX or non-AJAX, won't it produce problems?)
   ```php
          $loader = ($this->_name === 'html-header')
            ? 'var CRM = %s;' : 'CRM.$.extend(true, CRM, %s);';
          $js = sprintf($loader, json_encode($this->getSettings()));
          $html .= sprintf("<script type=\"text/javascript\">\n%s\n</script>\n", $js);
   ```

Reply:
@totten I don't think 2 would work - if a processor needs to inject code it won't necessarily have the chance to do so in html-header - because processor forms are often loaded when the processor is selected, or by webform

Reply:
@totten I've thought about getting rid of `|| !isAjaxMode()` because I also think it is superfluous. That might be a good goal to do before making other changes per this #18141, #18247.

In drupal 7 the html-header is *always* rendered if we are on a civicrm-ified page (eg. drupal_webform). But the renderSetting and definition of `var CRM` is *not* always called. It does get called if you specify a payment processor on the webform because `CRM.config.creditCardTypes` gets added which triggers the render settings code.

See: https://github.com/civicrm/civicrm-drupal/blob/7.x-master/civicrm.module#L156

As I understand it, if a payment processor is going to be loaded we *will* have `html-header`, it's just that `html-header` is static across payment processor changes so it can't be re-used by the processor (as you've cleverly reproduced in the timestamp example).

Reply:
I agree we should remove the magic region condition here. How about moving the condition to js:

https://github.com/civicrm/civicrm-core/pull/18262